src/mir.rs

```diff
@@ -675,6 +674,109 @@ impl MirFunction {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+pub struct MirForeignFunction {
+    name: Identifier,
+    foreign_name: Identifier,
+    args: Vec<(Identifier, MirType)>,
+    return_type: MirType,
+}
+
+impl MirForeignFunction {
+    pub fn new(
+        name: Identifier,
+        foreign_name: Identifier,
+        args: Vec<(Identifier, MirType)>,
+        return_type: MirType,
+    ) -> Self {
+        Self {
+            name,
+            foreign_name,
+            args,
+            return_type,
+        }
+    }
+
+    /// Declare this function to the compiler WITHOUT assembling it
+    fn declare(&self, funcs: &mut BTreeMap<Identifier, MirFunction>) -> Result<(), MirError> {
+        // Check if the function has already been declared
+        if funcs.contains_key(&self.name) {
+            Err(MirError::FunctionRedefined(self.get_name()))
+        } else {
+            // NOTE: because these declarations are only used for type checking and I dont want to rewrite them to much, we just define it as an oak function
+            funcs.insert(self.get_name(), MirFunction::new(
+                self.name.clone(),
+                self.args.clone(),
+                self.return_type.clone(),
+                vec![],
+            ));
+            Ok(())
+        }
+    }
+
+    fn assemble(
+        &self,
+        funcs: &BTreeMap<Identifier, MirFunction>,
+        structs: &BTreeMap<Identifier, MirStructure>,
+    ) -> Result<AsmForeignFunction, MirError> {
+        let mut mir_args = Vec::new();
+        let mut asm_args = Vec::new();
+        let mut vars = BTreeMap::new();
+        for (arg_name, arg_type) in &self.args {
+            // Add the arguments to the function's arguments, and to the map of variables.
+            // The map of variables are not used to determine the function's stack size
+            // at compile time, but are used to for resolving types.
+            asm_args.push((arg_name.clone(), arg_type.to_asm_type(structs)?));
+            vars.insert(arg_name.clone(), arg_type.clone());
+            mir_args.push(MirExpression::Variable(arg_name.clone()));
+        }
+
+        // NOTE: would prefer to do this at the ASM level, this way the compiler can ask the Target impl to do it with this as fallback
+        let statement = if self.return_type.get_size(structs)? != 0 {
+            MirStatement::Return(vec![MirExpression::ForeignCall(self.foreign_name.clone(), mir_args)])
+        } else {
+            MirStatement::Expression(MirExpression::ForeignCall(self.foreign_name.clone(), mir_args))
+        };
```

Comment:
This is the same logic as was used in the `Tir` level to generate the wrapper function for `extern fn` declarations